anitya/lib/backends/__init__.py

```diff
@@ -98,7 +105,13 @@ def expand_subdirs(self, url, glob_char="*"):
         text_regex = re.compile(r'^d.+\s(\S+)\s*$', re.I | re.M)
 
         if url_prefix != "":
-            dir_listing = self.call_url(url_prefix).text
+            resp = self.call_url(url_prefix)
+            # When FTP server is called, Response object is not created
+            # and we get binary string instead
+            try:
+                dir_listing = resp.text
+            except AttributeError:
+                dir_listing = resp.decode('utf-8')
```

Comment:
I just skimmed https://tools.ietf.org/html/rfc2640, but it sounds like there's no guarantees that the response will be UTF-8 encoded, and even if the spec said it MUST be, I'd be willing to bet there are some non-conforming servers out there. If it *should* only be ASCII/UTF-8 then this is fine, but this should handle an Unicode decoding exception and raise the appropriate Anitya exception explaining that the FTP service is misbehaving.

Reply:
I will add another try-except when trying to decode the binary string and if it's not UTF-8, I will just raise `AnityaPluginException`.
Maybe there is some way, how to get the encodings of the binary string and then use the appropriate decode method. This will be nice way how to do it.
I will check this.
It will be much easier, if `Requests` library could work with FTP.

Reply:
There is a requests-ftp extension, although I'm not sure if it works with current requests. It'd be nice to drop the FTP support, but I'm not sure if there's still a number of projects only available over FTP. I think just bailing is a reasonable approach. The specification indicates it should only ever be ASCII or UTF-8, but because FTP is so old I have no doubt there are servers out there not following the latest RFC about that.

Reply:
I looked at it, but it's only a template, not real library. It's only some basic functionality for future work.